### PR TITLE
providerserver: Implement native protocol version 5 support

### PR DESCRIPTION
**Pull request by bflad**
Reference: https://github.com/hashicorp/terraform-plugin-framework/issues/158

This enables provider developers to natively create and serve framework providers on protocol version 5, which is compatible with Terraform CLI 0.12 and later. For provider developers that are using terraform-plugin-mux to migrate from terraform-plugin-sdk to this framework, they can now omit the usage of the `tf6to5server` package by setting the `ServeOpts` type `ProtocolVersion` field to `5`.

Any protocol layer restrictions with the older protocol version 5 are raised as errors from the framework. For example, if `Attributes` is defined in a `tfsdk.Attribute` ("nested attributes"), the `GetProviderSchema` RPC response will include an error diagnostic.

This implementation is essentially a duplication of the `internal/fromproto6`, `internal/proto6server`, and `internal/toproto6` packages with renaming of 6 to 5 and any minor RPC naming differences to match the protocol (e.g. `ValidateDataResourceConfig` -> `ValidateDataSourceConfig`, `ValidateProviderConfig` -> `PrepareProviderConfig`, `ValidateResourceConfig` -> `ValidateResourceTypeConfig`).

### CODE REVIEW

**Review by bendbennett — APPROVED** [review]
I'm going to be completely honest, I only skimmed this owing to the size of the PR but it LGTM.

File: internal/fromproto5/doc.go [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Package fromproto5 contains functions to convert from protocol version 6
```
[the remark this answers is not in the document]
Comment (bflad) [reply internal/fromproto5/doc.go:1]: Ah yes, copypasta here and probably in the two other doc files, thanks for catching.

File: internal/proto5server/doc.go [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Package proto5server contains the provider server implementation compatible
+// with protocol version 6 (tfprotov5.ProviderServer).
```
Comment (bflad) [inline internal/proto5server/doc.go:2]:
```suggestion
// with protocol version 5 (tfprotov5.ProviderServer).
```

File: internal/toproto5/doc.go [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Package toproto5 contains functions to convert from framework types to
+// protocol version 6 (tfprotov5) types.
```
Comment (bflad) [inline internal/toproto5/doc.go:2]:
```suggestion
// protocol version 5 (tfprotov5) types.
```

**Pull request merged**